### PR TITLE
Auto migrating

### PR DESCRIPTION
**Pull request by matmair**
This is a follow up to #3215 but with smaller goals. This PR enables auto-running open migrations; makeing updates less error-prone.

<a href="https://gitpod.io/#https://github.com/inventree/InvenTree/pull/3741"></a>

### CODE REVIEW

File: InvenTree/InvenTree/tasks.py [hunk 1]
```diff
@@ -286,3 +293,60 @@ def send_email(subject, body, recipients, from_email=None, html_message=None):
...
+
+    # Set the application to maintenance mode - no access from now on.
+    logger.info('Going into maintenance')
+    set_maintenance_mode(True)
+    logger.info('Mainentance mode is on now')
+
+    # Check if we are worker - go kill all other workers then.
+    # Only the frontend workers run updates.
+
+    # TODO
```
Comment (matmair) [inline InvenTree/InvenTree/tasks.py:331]: Open TODO: Only run updates in the main UI - would be bad if multiple workers decide to update at the same time.
@SchrodingersGat do you think running in workers could also work?

File: InvenTree/config_template.yaml [hunk 2]
```diff
@@ -108,6 +108,10 @@ plugins_enabled: False
 #plugin_file: /path/to/plugins.txt
 #plugin_dir: /path/to/plugins/
 
+# Set this variable to True to enable auto-migrations and auto-updates (on bare-metal with git avai)
+# Alternatively, use the environment variable INVENTREE_AUTO_UPDATE
+auto_update: False
```
Comment (matmair) [inline InvenTree/config_template.yaml:113]: @SchrodingersGat should we enable this by default?

File: InvenTree/InvenTree/tasks.py [hunk 3]
```diff
@@ -286,3 +293,60 @@ def send_email(subject, body, recipients, from_email=None, html_message=None):
...
+    if True:
+        logger.info('Current process is a worker - shutting down cluster')
+
+    # Ok now we are ready to go ahead!
+    # To be sure we are in maintenance this is wrapped
+    with maintenance_mode_on():
+        logger.info('Starting migrations')
+        call_command('migrate', interactive=False)
```
Comment (matmair) [inline InvenTree/InvenTree/tasks.py:339]: We could also call the
- 1. `inv update` command
- 2. seperate django commands
To make the full run after an update

Comment (matmair) [reply InvenTree/InvenTree/tasks.py:339]: just migrations should be enough for now

File: InvenTree/InvenTree/tasks.py (hunk 1 above)
Comment (matmair) [reply InvenTree/InvenTree/tasks.py]: The worker is now detected by a cmd flag

**Comment by SchrodingersGat** [comment]
@matmair doesn't `invoke migrate` (which gets called by `invoke update`) already take care of this? What specifically is this alteration adding?

**Comment by matmair** [comment]
@SchrodingersGat it removes the need for the step ;-)
Especially for plugin installations, it makes it much harder to forget to run the update command and thus more accessible.

**Comment by SchrodingersGat** [comment]
Migration is only part of the update step though, it is also imperative that (at least) the user:

- Ensures required packages are up to date
- Ensures static files are up to date

So `invoke update` still needs to be run in the case of an update. I like containing all these steps into a single command, so that the sequencing is defined, and not up to the user

**Comment by matmair** [comment]
@SchrodingersGat working on the rest of the steps too but due to our unconventional way of translating JS I am running into some edge-cases with translating / collecting static

**Comment by matmair** [comment]
@SchrodingersGat @eeintech we should test-run this one a bit before releasing the next stable (at least 1-2 weeks)

**Comment by eeintech** [comment]
@matmair I can test drive this but not sure how, what do you expect us to run/keep an eye on?

**Comment by matmair** [comment]
@eeintech it should not break the install. I have some concerns with it freezing up during the worker shutdown/maintenance mode negotiation. If it does not do this we should be fine.

**Comment by eeintech** [comment]
@matmair so pull the PR and run `inv update`?

**Comment by matmair** [comment]
@eeintech a pull (and restart of workers and webserver) should be enough. The updates should run themself after that ;-)

**Comment by eeintech** [comment]
@matmair Sorry for the delay. I checked out your PR and started the server, this is what I got:

``` bash
$ inv server
Waiting for database...
Database connection sucessful!
  Running command git clone --filter=blob:none --quiet https://github.com/wolflu05/inventree-bulk-plugin /tmp/pip-install-clpgnqjn/inventree-bulk-plugin_39a780b633b64d3d9d53d1404a1a6ac3
09:35:25 [Q] INFO Enqueued 68435
09:35:26 [Q] INFO Enqueued 68436
  Running command git clone --filter=blob:none --quiet https://github.com/wolflu05/inventree-bulk-plugin /tmp/pip-install-7qhbbkv6/inventree-bulk-plugin_79e5c2ac77b5469b8ba04b4a9dfea268
09:35:38 [Q] INFO Enqueued 68437
09:35:39 [Q] INFO Enqueued 68438
Watching for file changes with StatReloader
Watching for file changes with StatReloader
Performing system checks...

System check identified no issues (1 silenced).

You have 1 unapplied migration(s). Your project may not work properly until you apply the migrations for app(s): InvenTree.
Run 'python manage.py migrate' to apply them.
January 06, 2023 - 09:35:39
Django version 3.2.16, using settings 'InvenTree.settings'
Starting development server at http://127.0.0.1:8000/
Quit the server with CONTROL-C.
```

So why didn't it apply the migration automatically?

**Comment by eeintech** [comment]
Side note: my local install takes a while now to start the server... about 30s! Is it really necessary to `git clone` plugins twice at each start? Why are the main reason it is so slow now?

**Comment by eeintech** [comment]
@wolflu05 I'm not sure why do you think I'm not on the right branch but I am. I don't have any issue switching branches and only following @matmair's instruction on how to test this :wink:

File: InvenTree/InvenTree/tasks.py [hunk 4]
```diff
@@ -447,3 +454,74 @@ def send_email(subject, body, recipients, from_email=None, html_message=None):
...
+
+
+@scheduled_task(ScheduledTask.DAILY)
+def check_for_migrations(worker: bool = True):
+    """Checks if migrations are needed.
+
+    If the setting auto_update is enabled we will start updateing.
+    """
+    # Test if auto-updates are enabled
+    if not get_setting('INVENTREE_AUTO_UPDATE', 'auto_update'):
```
Comment (wolflu05) [inline InvenTree/InvenTree/tasks.py:466]: @eeintech maybe this setting is of so its not working?

Comment (eeintech) [reply InvenTree/InvenTree/tasks.py:466]: If it is required, I don't find it in the "Server Settings" of the UI

Comment (wolflu05) [reply InvenTree/InvenTree/tasks.py:466]: I would say its required to set this via env `export INVENTREE_AUTO_UPDATE=True` or via config key `auto_update`, but @matmair could tell more.

Comment (matmair) [reply InvenTree/InvenTree/tasks.py:466]: It is intentionally not exposed as I am worried about crashes if it is changed while running.

**Comment by matmair** [comment]
@eeintech are you experiencing the same slow start on the current master?

**Comment by eeintech** [comment]
@matmair 

Running: `printf '%s %s\n' "$(date)" "Starting server" > server_start.log | inv server | while IFS= read -r line; do printf '%s %s\n' "$(date)" "$line"; done >>server_start.log`

Here's the output of with bulk plugin enabled:
``` bash
Mon 09 Jan 2023 02:34:38 PM EST Starting server
Mon 09 Jan 2023 02:34:45 PM EST Waiting for database...
Mon 09 Jan 2023 02:34:45 PM EST Database connection sucessful!
Mon 09 Jan 2023 02:34:46 PM EST   Running command git clone --filter=blob:none --quiet https://github.com/wolflu05/inventree-bulk-plugin /tmp/pip-install-fbtwgiw1/inventree-bulk-plugin_ec429efb492448b08d001b69e8d91041
Mon 09 Jan 2023 02:34:56 PM EST 14:34:56 [Q] INFO Enqueued 68465
Mon 09 Jan 2023 02:34:57 PM EST 14:34:57 [Q] INFO Enqueued 68466
Mon 09 Jan 2023 02:34:58 PM EST   Running command git clone --filter=blob:none --quiet https://github.com/wolflu05/inventree-bulk-plugin /tmp/pip-install-vapdc26y/inventree-bulk-plugin_d7ec31f5d09b4161a6ae5d27af5c96dc
Mon 09 Jan 2023 02:35:09 PM EST 14:35:09 [Q] INFO Enqueued 68467
Mon 09 Jan 2023 02:35:09 PM EST 14:35:09 [Q] INFO Enqueued 68468
Mon 09 Jan 2023 02:35:09 PM EST Watching for file changes with StatReloader
Mon 09 Jan 2023 02:35:09 PM EST Watching for file changes with StatReloader
Mon 09 Jan 2023 02:35:09 PM EST Performing system checks...
Mon 09 Jan 2023 02:35:09 PM EST 
Mon 09 Jan 2023 02:35:10 PM EST System check identified no issues (1 silenced).
Mon 09 Jan 2023 02:35:10 PM EST January 09, 2023 - 14:35:10
Mon 09 Jan 2023 02:35:10 PM EST Django version 3.2.16, using settings 'InvenTree.settings'
Mon 09 Jan 2023 02:35:10 PM EST Starting development server at http://127.0.0.1:8000/
Mon 09 Jan 2023 02:35:10 PM EST Quit the server with CONTROL-C.
```

With plugin disabled:
``` bash
Mon 09 Jan 2023 02:37:31 PM EST Starting server
Mon 09 Jan 2023 02:37:37 PM EST Waiting for database...
Mon 09 Jan 2023 02:37:37 PM EST Database connection sucessful!
Mon 09 Jan 2023 02:37:39 PM EST   Running command git clone --filter=blob:none --quiet https://github.com/wolflu05/inventree-bulk-plugin /tmp/pip-install-zz9hwtwd/inventree-bulk-plugin_8f462c995cc64c7a9dad228de32cb7c8
Mon 09 Jan 2023 02:37:49 PM EST 14:37:49 [Q] INFO Enqueued 68474
Mon 09 Jan 2023 02:37:51 PM EST   Running command git clone --filter=blob:none --quiet https://github.com/wolflu05/inventree-bulk-plugin /tmp/pip-install-rtw99032/inventree-bulk-plugin_54cf41d6ef684d32a2e159da777a392f
Mon 09 Jan 2023 02:38:00 PM EST 14:38:00 [Q] INFO Enqueued 68475
Mon 09 Jan 2023 02:38:01 PM EST Watching for file changes with StatReloader
Mon 09 Jan 2023 02:38:01 PM EST Watching for file changes with StatReloader
Mon 09 Jan 2023 02:38:01 PM EST Performing system checks...
Mon 09 Jan 2023 02:38:01 PM EST 
Mon 09 Jan 2023 02:38:01 PM EST System check identified no issues (1 silenced).
Mon 09 Jan 2023 02:38:01 PM EST January 09, 2023 - 14:38:01
Mon 09 Jan 2023 02:38:01 PM EST Django version 3.2.16, using settings 'InvenTree.settings'
Mon 09 Jan 2023 02:38:01 PM EST Starting development server at http://127.0.0.1:8000/
Mon 09 Jan 2023 02:38:01 PM EST Quit the server with CONTROL-C.
```

**Comment by eeintech** [comment]
> @eeintech is the startup speed difference significant between master and this branch?

No, as I said before it's been a few months now I have noticed that spinning off my development instance takes ~30s, it is not related to this branch.

> **Side note**: my local install takes a while now to start the server... about 30s! Is it really necessary to `git clone` plugins twice at each start? Why are the main reason it is so slow now?

**Comment by matmair** [comment]
@eeintech it should speed up a lot if you disable plugin checks on startup

**Comment by eeintech** [comment]
> @eeintech it should speed up a lot if you disable plugin checks on startup

Holly molly yes that's it:

``` bash
Mon 09 Jan 2023 03:49:17 PM EST Starting server
Mon 09 Jan 2023 03:49:23 PM EST Waiting for database...
Mon 09 Jan 2023 03:49:23 PM EST Database connection sucessful!
Mon 09 Jan 2023 03:49:25 PM EST 15:49:25 [Q] INFO Enqueued 68478
Mon 09 Jan 2023 03:49:27 PM EST 15:49:27 [Q] INFO Enqueued 68479
Mon 09 Jan 2023 03:49:27 PM EST Watching for file changes with StatReloader
Mon 09 Jan 2023 03:49:27 PM EST Watching for file changes with StatReloader
Mon 09 Jan 2023 03:49:27 PM EST Performing system checks...
Mon 09 Jan 2023 03:49:27 PM EST 
Mon 09 Jan 2023 03:49:27 PM EST System check identified no issues (1 silenced).
Mon 09 Jan 2023 03:49:28 PM EST January 09, 2023 - 15:49:28
Mon 09 Jan 2023 03:49:28 PM EST Django version 3.2.16, using settings 'InvenTree.settings'
Mon 09 Jan 2023 03:49:28 PM EST Starting development server at http://127.0.0.1:8000/
Mon 09 Jan 2023 03:49:28 PM EST Quit the server with CONTROL-C.
```

About 20s to check on plugins? :thinking:

**Comment by matmair** [comment]
@eeintech comment out the git line and install something else - maybe InvenTree-apprise . You will see the impact of shells is much smaller

**Comment by wolflu05** [comment]
> Git is not a delivery protocol. 

Ok. You got me. Didn't thought that this is needed. I'll try to publish my plugin on pypi to.

**Comment by matmair** [comment]
No pressure @wolflu05, it is a valid method that is supported by core - InvenTree normally does not start up that much anyway. Pypi has just a much more efficient version checking and delivery

**Comment by matmair** [comment]
@SchrodingersGat anything I should add/clarify here?

File: InvenTree/InvenTree/tasks.py [hunk 5]
```diff
@@ -447,3 +454,74 @@ def send_email(subject, body, recipients, from_email=None, html_message=None):
...
+
+    # Set the application to maintenance mode - no access from now on.
+    logger.info('Going into maintenance')
+    set_maintenance_mode(True)
+    logger.info('Mainentance mode is on now')
+
+    # Check if we are worker - go kill all other workers then.
+    # Only the frontend workers run updates.
+    if worker:
+        logger.info('Current process is a worker - shutting down cluster')
```
Comment (SchrodingersGat) [inline InvenTree/InvenTree/tasks.py:492]: @matmair should there be some more code here? Or just logging an info message?

**Comment by SchrodingersGat** [comment]
@matmair one thing I am unclear on - this will not fully handle an "update" automatically, right?

- `invoke install` will not be called (so if any requirements have changed, they will not be updated)
- translation files will not be updated
- static files will not be recompiled

Is this intended? Is the benefit of *only* doing migrations (which is handled by `invoke update`) enough? And, what errors might creep in if we automatically apply migrations but don't do the other required update steps?

**Comment by matmair** [comment]
@SchrodingersGat no, it only handles migrations. Because that is the step you can't really pre-run when using containers. The rest can be solved by mounting the pre-compiled static files and python packages dir. Pretty much the same as the container plugin install option.

**Comment by SchrodingersGat** [comment]
Does this also solve the problem of building the required tables when installing a new app plugin?

**Comment by matmair** [comment]
Yes, the next time any of the web workers start up again

**Comment by SchrodingersGat** [comment]
Currently, if you install a plugin which requires new db tables, what is the required series of steps to get those migrated?

**Comment by matmair** [comment]
Either run the whole update command or `inv migrate`. Neither can be done when deployed on a corporate cluster where you have no access to the cli.

**Comment by matmair** [comment]
Yes.
Fully integrated update might be difficult as we need to restart the frontend workers to load the new translation files.

**Comment by SchrodingersGat** [comment]
@matmair thanks for the hard work, happy to merge this in. Please provide some documentation for this feature:

- What it does
- Why a user / sysadmin would want to set it up, or in what circumstances
- How to use it
